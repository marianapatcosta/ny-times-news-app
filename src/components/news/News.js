import { defineComponent } from "vue";
import { formatDate } from "@/utils/utils";

export default defineComponent({
  name: "News",
  props: {
    newsInfo: {
      type: Object,
      required: true,
    },
  },
  data() {
    return {
      formatDate,
    };
  },
  methods: {
    handleOnClick() {
      this.$store.dispatch("setNewsInfo", this.newsInfo);
      this.$router.push(`/news/${this.newsInfo.id}`);
    },
  },
});
