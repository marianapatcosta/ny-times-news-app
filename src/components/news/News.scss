$space: 0.75rem;

.news {
  background-color: $color-primary;
  box-shadow: 0 0.125rem 0.5rem rgba($color-black, 0.7);
  cursor: pointer;
  display: flex;
  flex-direction: column;
  min-height: 20rem;
  overflow: hidden;
  width: 100%;
  max-width: 16rem;

  &:hover {
    transform: scale(1.01);
    -webkit-transform: scale(1.01);
    -moz-transform: scale(1.01);
    -ms-transform: scale(1.01);
    -o-transform: scale(1.01);
  }

  &__content {
    display: flex;
    flex-direction: column;
    flex: 1;
    padding: 1rem;
  }

  &__image {
    height: 11rem;
    object-fit: cover;
    object-position: 0 0;
  }

  &__image-placeholder {
    background-color: $color-secondary;
    opacity: 0.5;
  }

  &__header {
    font-size: 130%;
  }

  &__subheader {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-top: $space;
    font-size: 90%;
  }

  &__section-name {
    font-weight: 700;
  }

  &__date {
    font-style: italic;
  }

  &__details {
    margin: $space 0;
  }

  &__footer {
    font-size: 85%;
    margin-top: auto;
  }
}

@media only screen and (min-width: $xs) {
  .news {
    min-height: 23rem;
    max-width: 18.5rem;
  }
}
