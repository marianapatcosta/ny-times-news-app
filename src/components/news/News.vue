<template>
  <li
    class="news"
    data-cy="news"
    role="button"
    tabIndex="0"
    aria-label="Click to check news details."
    @click.prevent="handleOnClick"
    @keydown.enter="handleOnClick"
  >
    <img
      v-if="newsInfo.imageUrl"
      class="news__image"
      :src="`https://www.nytimes.com/${newsInfo.imageUrl}`"
      :alt="newsInfo.headline"
    />
    <div v-else class="news__image news__image-placeholder" />
    <div class="news__content">
      <h2 class="news__header" data-cy="news-header">
        {{ newsInfo.headline }}
      </h2>
      <div class="news__subheader">
        <time class="news__date" :datetime="newsInfo.pubDate">
          {{ formatDate(newsInfo.pubDate) }}
        </time>
        <div class="news__section-name">
          {{ newsInfo.sectionName }}
        </div>
      </div>

      <div class="news__main">
        <div class="news__details">{{ newsInfo.abstract }}</div>
      </div>

      <div class="news__footer" v-if="newsInfo.source">
        Source: {{ newsInfo.source }}
      </div>
    </div>
  </li>
</template>

<script src="./News.js"></script>

<style scoped lang="scss" src="./News.scss"></style>
