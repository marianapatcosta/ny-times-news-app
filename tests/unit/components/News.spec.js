import { createStore } from "vuex";
import { mount } from "@vue/test-utils";
import News from "@/components/news/News.vue";
import { newsInfoMock } from "../../mocks";

describe("News.vue", () => {
  let store;
  let wrapper;
  let mockRouter;
  const mockedSetNewsInfo = jest.fn();
  beforeEach(() => {
    const actions = {
      setNewsInfo: mockedSetNewsInfo,
    };

    store = createStore({
      actions,
    });

    mockRouter = {
      push: jest.fn(),
    };
    const mockRoute = {
      params: {
        id: newsInfoMock.id,
      },
    };
    wrapper = mount(News, {
      props: { newsInfo: newsInfoMock },
      global: {
        mocks: {
          $store: store,
          $route: mockRoute,
          $router: mockRouter,
        },
      },
    });
  });
  it("should render without errors and match snapshot renders correctly", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should render news info", () => {
    expect(wrapper.text()).toContain(newsInfoMock.headline);
  });

  it("should dispatch newsInfo to the store and redirect to 'News Details' page when item is clicked", async () => {
    await wrapper.trigger("click");
    expect(mockedSetNewsInfo).toHaveBeenCalled();
    expect(mockRouter.push).toHaveBeenCalledTimes(1);
    expect(mockRouter.push).toHaveBeenCalledWith(`/news/${newsInfoMock.id}`);
  });
});
