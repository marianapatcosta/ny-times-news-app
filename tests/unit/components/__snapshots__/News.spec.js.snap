// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`News.vue should render without errors and match snapshot renders correctly 1`] = `
<li
  aria-label="Click to check news details."
  class="news"
  data-cy="news"
  role="button"
  tabindex="0"
>
  <img
    alt="The headline"
    class="news__image"
    src="https://www.nytimes.com/images/2021/10/15/t-magazine/15tmag-harlemymca-slide-AEU3-copy/15tmag-harlemymca-slide-AEU3-copy-articleLarge.jpg"
  />
  <div
    class="news__content"
  >
    <h2
      class="news__header"
      data-cy="news-header"
    >
      The headline
    </h2>
    <div
      class="news__subheader"
    >
      <time
        class="news__date"
        datetime="2021-10-15T11:00:12+0000"
      >
        October 15, 2021
      </time>
      <div
        class="news__section-name"
      >
        T Magazine
      </div>
    </div>
    <div
      class="news__main"
    >
      <div
        class="news__details"
      >
        Lorem ipsum dolor sit amet. Non fugiat mollitia est nihil fugiat ea amet exercitationem. Et similique minus qui voluptas galisum et culpa tempora ea reprehenderit repudiandae.
      </div>
    </div>
    <div
      class="news__footer"
    >
       Source: The New York Times
    </div>
  </div>
</li>
`;
